src/System.Private.CoreLib/shared/System/Collections/Generic/Dictionary.cs

```diff
@@ -318,69 +336,83 @@ public virtual void GetObjectData(SerializationInfo info, StreamingContext conte
+                Debug.Assert(_entries != null, "expected entries to be != null");
                 IEqualityComparer<TKey>? comparer = _comparer;
                 if (comparer == null)
                 {
                     uint hashCode = (uint)key.GetHashCode();
                     // Value in _buckets is 1-based
-                    i = buckets[hashCode % (uint)buckets.Length] - 1;
+                    int i = buckets[hashCode % (uint)buckets.Length] - 1;
                     if (default(TKey)! != null) // TODO-NULLABLE: default(T) == null warning (https://github.com/dotnet/roslyn/issues/34757)
                     {
+                        Entry[]? entries = _entries;
+                        int collisionCount = 0;
                         // ValueType: Devirtualize with EqualityComparer<TValue>.Default intrinsic
                         while (true)
                         {
                             // Should be a while loop https://github.com/dotnet/coreclr/issues/15476
                             // Test in if to drop range check for following array access
-                            if ((uint)i >= (uint)entries.Length || (entries[i].hashCode == hashCode && EqualityComparer<TKey>.Default.Equals(entries[i].key, key)))
+                            if ((uint)i >= (uint)entries.Length)
                             {
-                                break;
+                                goto ReturnNotFound;
                             }
 
-                            i = entries[i].next;
+                            entry = ref entries[i];
+                            if (entry.hashCode == hashCode && EqualityComparer<TKey>.Default.Equals(entry.key, key))
+                            {
+                                goto ReturnFound;
+                            }
+
+                            i = entry.next;
                             if (collisionCount >= entries.Length)
                             {
                                 // The chain of entries forms a loop; which means a concurrent update has happened.
                                 // Break out of the loop and throw, rather than looping forever.
-                                ThrowHelper.ThrowInvalidOperationException_ConcurrentOperationsNotSupported();
+                                goto ConcurrentOperation;
                             }
                             collisionCount++;
                         }
                     }
                     else
                     {
+                        Entry[]? entries = _entries;
                         // Object type: Shared Generic, EqualityComparer<TValue>.Default won't devirtualize
                         // https://github.com/dotnet/coreclr/issues/17273
                         // So cache in a local rather than get EqualityComparer per loop iteration
                         EqualityComparer<TKey> defaultComparer = EqualityComparer<TKey>.Default;
```

Comment:
Why don't we cache this same `EqualityComparer<TKey>.Default;` value in the other branch of the if before the `while(true)` loop beginning at https://github.com/dotnet/coreclr/pull/27149/files#diff-c925cc7aff8edb06fbe36ff38cfbfc57R362 since it gets used at least once at https://github.com/dotnet/coreclr/pull/27149/files#diff-c925cc7aff8edb06fbe36ff38cfbfc57R372 inside the loop (unless `i` starts out `>= buckets.Length`, which should have triggered a divide by zero on the `hashCode` modulus operation in line 356)?

Reply:
`EqualityComparer<T>.Default` is a Jit intrinsic https://github.com/dotnet/coreclr/pull/14125 and it methods become direct calls and it evaporates; so for structs `default(TKey) != null` it devirtualizes directly to equality and hashcode. e.g. for `int` GeyHashCode becomes the same value and equality is just a `cmp` when it becomes asm.

However; for class types the generic method is shared between all types (for structs a new method is specialised for each type); this means that `EqualityComparer<T>.Default` can't devirtualize in the asm as it doesn't know what type `TKey` is; so we have to pay a small runtime lookup cost to get the actual Comparer to use. So on that branch we hoist getting the comparer outside the loop so we only pay the lookup cost once.

There's an issue for the Jit being smarter and hoisting the comparer lookup itself outside of the loop in shared generics https://github.com/dotnet/coreclr/issues/17273; but it doesn't do it yet.

Reply:
Normally you'd just use `EqualityComparer<T>.Default` directly or more simply constrain the type to `IEquatable<TKey>`; however `Dictionary<TKey,TValue>` isn't constrained and allows custom comparerer, while being a fairly fundamental backbone type and used very heavily by everything. So in the performance monitoring a regression was picked up and the two paths were split to maintain the high performance: https://github.com/dotnet/coreclr/pull/17285

Reply:
thanks for the clear explaining :)